classes/ActionScheduler_AbstractAction.php

```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * Class ActionScheduler_AbstractAction
+ *
+ * Define shared methods and data for all action types.
+ *
+ * @since 1.6.0
+ */
+abstract class ActionScheduler_AbstractAction implements ActionScheduler_Interface_Action, ActionScheduler_Interface_Scheduled {
+
+	/** @var string */
+	protected $hook = '';
+
+	/** @var array */
+	protected $args = array();
+
+	/** @var ActionScheduler_Schedule */
+	protected $schedule = NULL;
+
+	/** @var string */
+	protected $group = '';
+
+	/** @var bool */
+	protected $is_finished = false;
+
+	public function execute() {
+		return do_action_ref_array( $this->get_hook(), $this->get_args() );
```

Comment:
> `do_action_ref_array()` doesn't actually return anything, so we could eliminate the preceding return here.

I've changed that with SHA: cd6c1ac It was carried over from `ActionScheduler_Action`, but seeing as we're introducing the changeset, now seems as good a time as any to fix that up.

> should the call to `do_action_ref_array()` be preceded with `if ( ! $this->is_finished )`, followed by setting `$this->is_finished` to true? Or have I missed where the `$is_finished` property is used elsewhere?

I toyed with that idea, and while it's correct, it would be a breaking change. Previously, calling `ActionScheduler_Action->execute()` twice consecutively would execute it twice. If we introduced that change, it 
would only be executed once (there are valid reasons why someone might do that, like testing an action for fatal errors before rolling back, albeit unlikely to actually be being done by anyone).

I also toyed with deprecating `execute()` and instead using a new `run()` method which could implement that, and/or update the action's internal status once it had run rather than it relying on `ActionScheduler_QueueRunner::process_action()` to call `ActionScheduler_QueueRunner->store->mark_complete()` to do it.

Really though, `is_finished()` and the `$is_finished` property are only used when saving an action's status now so it could also theoretically be removed entirely and the status property used instead.

For now, I've kept it simple, but if we're going to change AS to be v2.0, we could go for more extreme breaking changes like this.